hoverkite-protocol/src/response.rs

```diff
@@ -89,178 +130,157 @@ impl Report {
-                Some(Self {
-                    side,
-                    response: Response::Report(Report::Position(position)),
-                })
-            } else {
-                None
-            }),
-            Some(b'B') | Some(b'b') => Ok(if buffer.len() >= 7 {
-                let side = if buffer.pop_front().unwrap() == b'B' {
-                    side
-                } else {
-                    side.opposite()
-                };
-                let bytes: Vec<u8> = buffer.drain(0..6).collect();
-                let battery_voltage = u16::from_le_bytes(bytes[0..2].try_into().unwrap());
-                let backup_battery_voltage = u16::from_le_bytes(bytes[2..4].try_into().unwrap());
-                let motor_current = u16::from_le_bytes(bytes[4..6].try_into().unwrap());
-                Some(Self {
-                    side,
-                    response: Response::Report(Report::BatteryReadings {
-                        battery_voltage,
-                        backup_battery_voltage,
-                        motor_current,
-                    }),
-                })
-            } else {
-                None
-            }),
-            Some(b'C') | Some(b'c') => Ok(if buffer.len() >= 2 {
-                let side = if buffer.pop_front().unwrap() == b'C' {
-                    side
-                } else {
-                    side.opposite()
-                };
-                let byte = buffer.pop_front().unwrap();
-                let charger_connected = match byte {
-                    b'0' => false,
-                    b'1' => true,
-                    _ => return Err(UnexpectedResponse(byte)),
-                };
-                Some(Self {
-                    side,
-                    response: Response::Report(Report::ChargeState { charger_connected }),
-                })
-            } else {
-                None
-            }),
-            Some(r) => {
-                buffer.pop_front();
-                Err(UnexpectedResponse(r))
-            }
-            None => Ok(None),
+    #[cfg(feature = "std")]
+    pub fn write_to_std(&self, writer: impl std::io::Write) -> std::io::Result<()> {
+        self.write_to(&mut WriteCompat(writer))
+    }
+
+    pub fn write_to<W>(&self, writer: &mut W) -> Result<(), W::Error>
+    where
+        W: embedded_hal::blocking::serial::Write<u8>,
+    {
+        writer.bwrite_all(&[self.side.to_byte()])?;
+        self.response.write_to(writer)
+    }
+
+    pub fn parse(buffer: &[u8]) -> nb::Result<Self, ProtocolError> {
+        if let [side, ref rest @ ..] = *buffer {
+            Ok(SideResponse {
+                side: Side::parse(side)?,
+                response: Response::parse(rest)?,
+            })
+        } else {
+            Err(WouldBlock)
         }
     }
 }
 
-#[cfg(feature = "std")]
 #[cfg(test)]
 mod tests {
     use super::*;
     use test_case::test_case;
 
+    mod log {
+        use super::super::*;
+        use test_case::test_case;
```

Comment:
Nit: You could just `use super::*` and get both of these.

Reply:
there is a `test_case` macro in std::prelude::v1 behind the custom_test_frameworks feature flag, and glob imports have the same precedence as prelude imports, so you get this warning:

```
error[E0659]: `test_case` is ambiguous (glob import vs any other name from outer scope during import/macro resolution)
   --> src/response.rs:179:11
    |
179 |         #[test_case("$" ; "width 1")]
    |           ^^^^^^^^^ ambiguous name
    |
note: `test_case` could refer to the attribute macro imported here
   --> src/response.rs:177:13
    |
177 |         use super::*;
    |             ^^^^^^^^
    = help: consider adding an explicit import of `test_case` to disambiguate
    = help: or use `self::test_case` to refer to this attribute macro unambiguously
note: `test_case` could also refer to the attribute macro defined here
   --> /Users/david/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/prelude/mod.rs:105:13
    |
105 |     pub use super::v1::*;
    |             ^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0659`.
error: could not compile `hoverkite-protocol`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
```